File: src/agents/engineeringExecution/shell.ts
import { spawn } from 'node:child_process';

export interface CommandResult {
  command: string;
  code: number;
  output: string;
}

export function run(cmd: string[], opts: { cwd?: string; timeoutMs?: number; input?: string } = {}): Promise<CommandResult> {
  const timeoutMs = opts.timeoutMs ?? 600_000;
  return new Promise((resolve) => {
    const [bin, ...args] = cmd;
    const child = spawn(bin!, args, { cwd: opts.cwd });
    let output = '';
    let settled = false;

    if (opts.input !== undefined) child.stdin?.end(opts.input);
    else child.stdin?.end();

    const timer = setTimeout(() => {
      if (settled) return;
      settled = true;
      child.kill('SIGKILL');
      resolve({ command: cmd.join(' '), code: 124, output: output + `\n[timed out after ${timeoutMs}ms]` });
    }, timeoutMs);

    child.stdout?.on('data', (d) => (output += d.toString()));
    child.stderr?.on('data', (d) => (output += d.toString()));
    child.on('close', (code) => {
      if (settled) return;
      settled = true;
      clearTimeout(timer);
      resolve({ command: cmd.join(' '), code: code ?? 1, output });
    });
    child.on('error', (err) => {
      if (settled) return;
      settled = true;
      clearTimeout(timer);
      resolve({ command: cmd.join(' '), code: 1, output: output + `\n${String(err)}` });
    });
  });
}

export async function runOrThrow(cmd: string[], opts: { cwd?: string; timeoutMs?: number; input?: string } = {}): Promise<CommandResult> {
  const result = await run(cmd, opts);
  if (result.code !== 0) {
    throw new Error(`command failed (${result.command}):\n${result.output.slice(-2000)}`);
  }
  return result;
}

File: legacy-python/work_agent/execution/shell.py
from __future__ import annotations

import subprocess
from dataclasses import dataclass


@dataclass
class CommandResult:
    command: str
    returncode: int
    output: str  # combined stdout+stderr

    @property
    def ok(self) -> bool:
        return self.returncode == 0


def run(cmd: list[str], cwd: str | None = None, timeout: int = 600, env: dict | None = None) -> CommandResult:
    try:
        proc = subprocess.run(
            cmd,
            cwd=cwd,
            timeout=timeout,
            env=env,
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            text=True,
        )
        return CommandResult(command=" ".join(cmd), returncode=proc.returncode, output=proc.stdout)
    except subprocess.TimeoutExpired as exc:
        output = (exc.stdout or "") if isinstance(exc.stdout, str) else ""
        return CommandResult(command=" ".join(cmd), returncode=124, output=output + f"\n[timed out after {timeout}s]")


def run_or_raise(cmd: list[str], cwd: str | None = None, timeout: int = 600) -> CommandResult:
    result = run(cmd, cwd=cwd, timeout=timeout)
    if not result.ok:
        raise RuntimeError(f"command failed ({result.command}):\n{result.output[-2000:]}")
    return result
